chimedb/dataset/get.py

```diff
@@ -21,42 +24,300 @@
+        This does not insert a new row into the database.
+
+        Parameter
+        ----------
+        ds_id
+            Dataset ID.
+
+        Returns
+        -------
+        Dataset
+            The requested dataset.
+        """
+        if ds_id not in _dataset_cache:
+            try:
+                _dataset_cache[ds_id] = Dataset.get(Dataset.id == ds_id)
+            except orm.Dataset.DoesNotExist:
+                _logger.warning("Could not find dataset {}.".format(ds_id))
+                return None
+        return _dataset_cache[ds_id]
+
+    @property
+    def type(self):
+        """Get the type of the attached dataset state."""
+        if self._type is None:
+            return self.dataset_state.type
+        return self._type
+
+    @property
+    def base_dataset(self):
+        """Get the base dataset."""
+        if not self._base_dataset:
+            self._base_dataset = self.base_dset
+        return self._base_dataset
+
+    @property
+    def dataset_state(self):
+        """Get the dataset state."""
+        # get the raw ID of the state (foreign key)
+        orm_base_class = super()
+        state_id = orm_base_class.state_id
+
+        if state_id not in _state_cache:
+            _state_cache[state_id] = orm_base_class.state
+        return _state_cache[state_id]
+
+    def __repr__(self):
+        if self._type:
+            return "<get.Dataset[{}]: {}>".format(self.type, self.id)
+        else:
+            return "<get.Dataset: {}>".format(self.id)
+
+    def closest_ancestor_of_type(self, type_):
+        """
+        Get the closest ancestorial dataset of the given type.
+
+        Parameters
+        ----------
+        type_ : str or :class:`chimedb.dataset.orm.DatasetStateType`
+
+        Raises
+        ------
+        :class:`chimedb.dataset.orm.DatasetStateType.DoesNotExist`
+            If `type_` is not an existing dataset state type.
+
+        Returns
+        -------
+        :class:`chimedb.dataset.orm.Dataset`
+            The closest ancestor of the given type. Returns `None` if no ancestor of the
+            given type found.
+        """
+        if isinstance(type_, str):
+            type_ = DatasetStateType.from_name(name=type_)
+            if type_ is None:
+                raise NotFoundError("{} is not a known DatasetStateType".format(type_))
+            type_ = type_.name
+        d = self
+
+        while d:
+            if d.type == type_:
+                return d
+            d = d.base_dataset
+
+        raise NotFoundError(
+            "No ancestor of type {} found for Dataset {}".format(type_, self.__repr__())
+        )
+
+
+def index():
+    """Pre-fetch and cache all Dataset(State(Type))s."""
+    query = (
+        orm.Dataset.select(orm.Dataset, orm.DatasetStateType.name.alias("_type"))
+        .join(orm.DatasetState)
+        .join(orm.DatasetStateType)
+    )
+
+    # Create a mapping from dataset ID to
+    dsdict = {d.id: d for d in query.objects(Dataset)}
```

Comment:
This query takes an amazingly long amount of time. In order the queries in here take: ~23s, 0.25s and 0.01s. It's odd because while there are 7x more rows in the Dataset table, the amount in each row is pretty similar, but it takes 100x longer.

Actually I tell a lie, in those numbers I modified the query to skip the joins (which are unnecessary except to get _type). However, that doesn't make any difference. I'm very puzzled by which it takes so much longer.

Perhaps the database schema is odd? This might be worth looking into a bit more, the fact the primary key is the hash, but things will have been inserted in there in a random order (as far as the primary key/hash is concerned) maybe means that data is non-contiguous and needs to get sorted on the way out? And maybe @ketiltrout has some ideas

Regardless 23s to fetch 140k fairly small rows seems a long time.